org.eclipse.jdt.core.manipulation/core extension/org/eclipse/jdt/internal/corext/fix/helper/WhileToForEach.java

```diff
@@ -0,0 +1,275 @@
+/*******************************************************************************
+ * Copyright (c) 2021, 2022 Carsten Hammer.
+ *
+ * This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Carsten Hammer
+ *******************************************************************************/
```

Comment:
The Copyright header is just copied from my sandbox project. Please tell me how to adjust or do it yourself. As minimal solution you should add yourself in all files changed or new.

Reply:
Feel free to push into the PR.  I will start working on the new test scenarios.  I think the complex one with if statements should be ignored.  The iterator() call should be in the same block as the while loop that uses it.

Reply:
Hi Carsten, I decided to remove the extra groups.  I was trying to solve a bug that I know believe is in the rewrite code, but I ended up just leaving them out as I think it is not needed and adds complexity.  I added support for checking that the loop var is used and adhering to the sub-preference which is necessary because the end-loop will show a warning/error and the user needs to decide whether it is ok.  I added a couple of new nested tests.  I also added a check where the code just calls the it.next() method without doing anything with it.  This type of call can't be converted to the loop var name without causing an error (e.g. it.next(); -> string;)  I did some formatting changes and converting some variables to camel-case.

FWIW: the failure occurs when I have a nested while and after the inner while, there is only comments.  It could be any while but this is how I found it.  I tried a variety of attempts to counter it (e.g. converting inner loop first), but it appears it is an error somwhere in the rewrite code (probably doesn't copy or move the comment properly).

    void m(List<String> strings,List<String> strings2) {
        Collections.reverse(strings);
        Iterator it = strings.iterator();
        while (it.hasNext()) {
        	it.next();
            Iterator it2 = strings2.iterator();
            while (it2.hasNext()) {
                String s2 = (String) it2.next();
                System.out.println(s2);
            }
            // this extraneous comment causes a range error to occur
        }
    }